model/inference.py

```diff
@@ -0,0 +1,171 @@
+import pickle
+import re
+
+import networkx as nx
+import numpy as np
+import pandas as pd
+import torch
+
+from .sobog import SOBOG
+
+_URL_PATTERN = r"[(http(s)?):\/\/(www\.)?a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,6}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)"  # noqa: E501
+
+
+class Inference:
+    def __init__(self):
+        model_path = "model/model.pt"
+        tfidf_path = "vect/vectorizer.pk"
+        user_mean_path = "ckpts/mean.csv"
+        user_std_path = "ckpts/std.csv"
+        self.user_columns = [
+            "statuses_count",
+            "followers_count",
+            "friends_count",
+            "favourites_count",
+            "listed_count",
+            "default_profile",
+            "default_profile_image",
+            "protected",
+            "verified",
+            "updated",
+            "created_at",
+            "name",
+            "screen_name",
+            "description",
+        ]
+        model_dict = {
+            "n_user_features": 20,
+            "d_user_embed": 20,
+            "n_post_features": 5000,
+            "d_post_embed": 64,
+            "n_gat_layers": 3,
+            "d_cls": 32,
+            "n_cls_layer": 2,
+        }
+
+        self.model = self.load_model(model_path, model_dict)
+        self.vectorizer = self.load_vectorizer(tfidf_path)
+        self.user_mean = pd.read_csv(user_mean_path)
+        self.user_std = pd.read_csv(user_std_path)
+
+    def create_user_dataframe(self, user):
+        return pd.DataFrame([user], columns=self.user_columns)
+
+    def preprocessing_user(self, user_df, user_mean, user_std):
+        if "updated" in user_df.columns:
+            age = (
+                pd.to_datetime(user_df.loc[:, "updated"])
+                - pd.to_datetime(user_df.loc[:, "created_at"]).dt.tz_localize(None)
+            ) / np.timedelta64(1, "Y")
+        else:
+            age = (
+                pd.to_datetime(pd.to_datetime("today"))
+                - pd.to_datetime(user_df.loc[:, "created_at"]).dt.tz_localize(None)
+            ) / np.timedelta64(1, "Y")
+        user_df["tweet_freq"] = user_df["statuses_count"] / age
+        user_df["followers_growth_rate"] = user_df["followers_count"] / age
+        user_df["friends_growth_rate"] = user_df["friends_count"] / age
+        user_df["favourites_growth_rate"] = user_df["favourites_count"] / age
+        user_df["listed_growth_rate"] = user_df["listed_count"] / age
+        user_df["followers_friends_ratio"] = user_df["followers_count"] / np.maximum(
+            user_df["friends_count"], 1
+        )
+        user_df["screen_name_length"] = user_df["screen_name"].str.len()
+        user_df["num_digits_in_screen_name"] = user_df["screen_name"].str.count(r"\d")
+        user_df["name_length"] = user_df["name"].str.len()
+        user_df["num_digits_in_name"] = user_df["name"].str.count(r"\d")
+        user_df["description_length"] = user_df["description"].str.len()
+        user_df.replace({False: 0, True: 1}, inplace=True)
```

Comment:
Didn't see this in original code, what's this for?
`user_df.replace({False: 0, True: 1}, inplace=True)`

Reply:
Preprocess boolean into integer

Reply:
Dint see that in origin? U sure this doesnt change anything?

Reply:
Because in sample data, they specify boolean as `0.0` and `1.0`